File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.py
from collections import deque
def solution(players, m, k):
    ans = 0
    # 하루동안 최소 몇 번 증설해야하는지
    # players[i] = i~i시 사이의 이용자 수
    # m명 = 1대 서버 감당 가능
    # 한 번 증설한 서버는 k 시간 동안 운영
    # stack으로 stack[0]에 가장 빨리 증설한 서버 종료 시각 저장
    # 매번 시각과 종료 시각을 비교하기
    stack = deque()
    for i,p in enumerate(players):
        # 플레이하는 사람 없으면 pass
        if p == 0: continue
        st = i
        et = i+1
        while stack and stack[0] < et:
            stack.popleft()
        server = len(stack)
        # 현재 감당 가능 인원보다 더 많은 이가 플레이할 경우
        if server * m < p:
            total = p // m
            total -= server
            for _ in range(total):
                stack.append(st+k)
            ans += total
    return ans

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.kt
import java.util.*;
class Solution {
    fun solution(players: IntArray, m: Int, k: Int): Int {
        var ans = 0
        val q = ArrayDeque<Int>()
        for ((i,p) in players.withIndex()) {
            val st = i
            val et = i+1
            while (!q.isEmpty() && q.peekFirst() < et) {
                q.pollFirst()
            }
            
            val server = q.size
            if (server * m < p) {
                var t = p / m
                t -= server
                repeat(t) {
                    q.addLast(st+k)
                }
                ans += t
            }
        }
        return ans
    }
}